graph/graphs/gen: Added breadth-first tree constructor

mattConn opened the pull request:
This adds a tree constructor per #1582. I made the decision to have the tree build breadth-first based on a specified uniform fanout, and allowed for it to be partially full, as in missing leaves. My method for calculating depth based on number of nodes required the use of a logarithm and so `math` is imported.

kortschak commented:
This needs some tests. When you add them, please change the commit message to be "graph/graphs/gen: add ..."; the convention is that the part of the commit message after the colon is a sentence fragment (so lower case) that completes the sentence "This change modifies Gonum to [add breadth-first tree constructor]."

mattConn commented:
Apologies, I completely forgot them. I'll add tests and examples.
> When you add them, please change the commit message ...

Will do.

mattConn commented:
I've modified the tree function significantly. It no longer depends on `math` to find tree depth and simply makes edges instead of calling `Star`. I've also added several tests and an example.

I originally intended to use `Star` but this seemed more practical, one reason being that a single call of `check` can be made versus one followed by many with multiple calls to `Star`. The other reason is that each call to `Star` requires its own `IDer` for its leaves, which would have to be made from the original `IDer` passed to `Tree`.

mattConn commented on graph/graphs/gen/gen.go:
```diff
@@ -175,6 +175,31 @@ func Wheel(dst NodeIDGraphBuilder, center int64, cycle IDer) {
+// If dst is a directed graph, edges are directed from the root to the leaves.
+// The tree is allowed to have missing leaves. If any ID appears more than
+// once in nodes, Tree will panic.
+func Tree(dst NodeIDGraphBuilder, n int, nodes IDer) {
```
A panic sounds appropriate. I'll update the "bad fanout" test to account for that.

By "missing leaves" I meant that the "last row" of the tree may be missing nodes if the length of the nodes supplied to the function doesn't equal `\sum_{i=0}^h n^i`. I took leaves in a tree to mean nodes with a single edge.

kortschak replied:
Thanks. Yeah, "leaves" is understood, it's the missingness that was ambiguous. Text along the lines of what I wrote (either expression) would cover that.

mattConn replied:
Done.

mattConn commented on graph/graphs/gen/gen.go:
```diff
@@ -175,6 +175,31 @@ func Wheel(dst NodeIDGraphBuilder, center int64, cycle IDer) {
+// The tree is allowed to have missing leaves. If any ID appears more than
+// once in nodes, Tree will panic.
+func Tree(dst NodeIDGraphBuilder, n int, nodes IDer) {
+	if n >= nodes.Len() || n <= 0 {
+		return
+	}
```
Done.

mattConn commented on graph/graphs/gen/gen_example_test.go:
```diff
@@ -200,3 +200,49 @@ func ExampleComplete_undirectedSet() {
+	// 	12;
+	// 	13;
+	// 	14;
+
```
Done.

mattConn commented on graph/graphs/gen/gen_example_test.go:
```diff
@@ -200,3 +200,49 @@ func ExampleComplete_undirectedSet() {
+	fmt.Printf("%s\n", b)
+
+	// Output:
+	// `strict graph full_binary_tree_undirected {
```
Done.

mattConn commented on graph/graphs/gen/gen_example_test.go:
```diff
@@ -200,3 +200,49 @@ func ExampleComplete_undirectedSet() {
+	// 	5 -- 12;
+	// 	6 -- 13;
+	// 	6 -- 14;
+	//    }`,
```
Done.

mattConn commented on graph/graphs/gen/gen_example_test.go:
```diff
@@ -200,3 +200,49 @@ func ExampleComplete_undirectedSet() {
+
+func ExampleTree_undirectedRange() {
+	dst := simple.NewUndirectedGraph()
+	gen.Tree(dst, 3, gen.IDRange{First: 0, Last: 14})
```
I may have added this example in haste, apologies. Done.

kortschak commented on graph/graphs/gen/gen.go:
```diff
@@ -175,6 +175,33 @@ func Wheel(dst NodeIDGraphBuilder, center int64, cycle IDer) {
+
+	j := 0
+	for i := 0; j < nodes.Len(); i++ {
+		u, _ := dst.NodeWithID(nodes.ID(i))
```
I think this is not right for the case of n=1, |nodes|=1; there needs to be a special case for that so a singleton tree is actually added. Sorry I missed that before.

mattConn replied:
Would a check of length of nodes at the top of `Tree` be appropriate? This check would only occur once versus the check that would be necessary for adding a node before setting each edge. Something like the following code:
```
func Tree(dst NodeIDGraphBuilder, n int, nodes IDer) {
	if nodes.Len() == 1 {
		if u, new := dst.NodeWithID(nodes.ID(0)); new {
			dst.AddNode(u)
		}
		return
	}
.
.
.
```

kortschak replied:
Yep, but also with a fast return for no nodes since we're here.

```
switch nodes.Len() {
case 0:
	return
case 1:
	if u, new := dst.NodeWithID(nodes.ID(0)); new {
		dst.AddNode(u)
	}
	return
}
```

mattConn replied:
Done.

kortschak replied:
And so it goes... We need an update of the docs now (sorry). We need to say that as special cases, |nodes| == 0 is a no-op and |nodes| == 1 will result a singleton graph (if previously empty). Currently the docs say that both of these will panic. Something like:

```
// If the number of nodes is greater than one, n must be non-zero and
// less than the number of nodes unless, otherwise Tree will panic.
```

mattConn replied:
Not a problem. It runs a little long, but how does this look:

```
// Tree constructs an n-ary tree breadth-first in dst with the given fan-out, n.
// If the number of nodes does not equal \sum_{i=0}^h n^i, where h is an integer
// indicating the height of the tree, a partial tree will be constructed with not
// all nodes having zero or n children, and not all leaves being h from the root.
// If the number of nodes is zero, Tree will do nothing.
// If the number of nodes is one, Tree will only add that node to dst.
// If the number of nodes is greater than one, n must be non-zero and less than the
// number of nodes, otherwise Tree will panic.
// If dst is a directed graph, edges are directed from the root to the
// leaves. If any ID appears more than once in nodes, Tree will panic.
```

kortschak replied:
I think that either bullet point special cases (like in stdlib's math) or just the two lines I have above (in terms of detail - if not exact wording - the special cases are only to avoid confusion with the panic, they do what you would expect so I don't think they need to be fully spelled out).

mattConn replied:
Done.

mattConn commented:
Please take a look.

mattConn commented:
In addition to the nodes length switch, I also updated the empty graph test and added a test for singleton trees.

kortschak reviewed (dismissed):
Thanks

mattConn commented:
Thanks for working with me on this.

kortschak commented:
A pleasure. You can go ahead and rebase this onto master now.

kortschak commented:
Oh, sorry. That's not what I meant. You can click the "rebase and merge" button.
